File: Python/1067-sort_with_wap(0,i).py


def swap_count(nums):
    num_to_idx = {}
    for idx, num in enumerate(nums):
        num = int(num)
        if idx != num or idx == 0:
            num_to_idx[num] = idx
    count = 0
    while num_to_idx:
        idx = num_to_idx[0]
        if idx != 0:
            num_to_idx[0] = num_to_idx[idx]
            del num_to_idx[idx]
            count += 1
        elif idx == 0 and len(num_to_idx) == 1:
            del num_to_idx[idx]
        else:
            for key in num_to_idx:
                if key != 0:
                    num_to_idx[0] = num_to_idx[key]
                    num_to_idx[key] = 0
                    count += 1
                    break
    return count


if __name__ == '__main__':
    _ = input()
    nums = [int(i) for i in input().split()]
    print(swap_count(nums))

File: C++/1067-sort_with_wap(0,i).cc
#include <iostream>
#include <map>


int main() {

    int n_num;
    std::cin >> n_num;
    std::map<int, int> num_to_idx;
    for (int i=0; i<n_num; ++i) {
        int num;
        std::cin >> num;
        if (i != num or i == 0) {
            num_to_idx[num] = i;
        }
    }
    int count = 0;
    while (!num_to_idx.empty()) {
        int idx = num_to_idx[0];
        if (idx != 0) {
            num_to_idx[0] = num_to_idx[idx];
            num_to_idx.erase(idx);
            ++count;
        } else if (idx == 0 && num_to_idx.size() == 1) {
            num_to_idx.erase(idx);
        } else {
            for (auto &kv : num_to_idx) {
                if (kv.first != 0) {
                    num_to_idx[0] = num_to_idx[kv.first];
                    num_to_idx[kv.first] = 0;
                    ++count;
                    break;
                }
            }
        }
    }
    std::cout << count << '\n';

    return 0;
}
